File: app/models/spot.py
from .db import db


class Spot(db.Model):
    __tablename__ = "spots"

    id = db.Column(db.Integer, primary_key=True)
    type = db.Column(db.String(255), nullable=False)
    name = db.Column(db.String(500), nullable=False)
    address = db.Column(db.String(1000), nullable=False)
    longitude = db.Column(db.Numeric(10, 7), nullable=False)
    latitude = db.Column(db.Numeric(10, 7), nullable=False)
    pets = db.Column(db.Boolean, nullable=False)
    total_occupancy = db.Column(db.Integer, nullable=False)
    total_bedrooms = db.Column(db.Integer, nullable=False)
    total_bathrooms = db.Column(db.Integer, nullable=False)
    description = db.Column(db.Text, nullable=False)
    has_wifi = db.Column(db.Boolean, nullable=False)
    has_tv = db.Column(db.Boolean, nullable=False)
    has_ac = db.Column(db.Boolean, nullable=False)
    price_per_day = db.Column(db.Numeric(9, 2), nullable=False)
    host_id = db.Column(db.Integer, db.ForeignKey("users.id"), nullable=False)
    created_at = db.Column(db.DateTime, default=db.func.now(), nullable=False)
    updated_at = db.Column(db.DateTime, default=db.func.now(), nullable=False)

    user = db.relationship("User", back_populates="spots")
    images = db.relationship(
        "Image", back_populates="spot", cascade="all, delete-orphan"
    )
    reservations = db.relationship(
        "Reservation", back_populates="spot", cascade="all, delete-orphan"
    )
    reviews = db.relationship(
        "Review", back_populates="spot", cascade="all, delete-orphan"
    )

    def get_reviews(self):
        spots_review = {}
        reviews = self.reviews

        for review in reviews:
            spots_review[review.id] = review.to_dict()

        return spots_review

    def to_dict(self):
        return {
            "id": self.id,
            "type": self.type,
            "name": self.name,
            "user_id": str(self.user.id),
            "pets_allowed": str(self.pets),
            "total_occupancy": str(self.total_occupancy),
            "total_bedrooms": str(self.total_bedrooms),
            "total_bathrooms": str(self.total_bathrooms),
            "description": str(self.description),
            "has_wifi": str(self.has_wifi),
            "has_tv": str(self.has_tv),
            "has_ac": str(self.has_ac),
            "price_per_day": str(self.price_per_day),
            "host_id": str(self.host_id),
            "images": [image.image_url for image in self.images],
            "host_name": self.user.name,
            "address": self.address,
            "longitude": float(self.longitude),
            "latitude": float(self.latitude),
            "reviews": self.get_reviews(),
        }

File: react-app/src/components/Spot/Spot.js
import { useEffect, useState } from "react";
import { useSelector } from "react-redux";
import { useParams } from "react-router-dom";
import NavBar from "../NavBar/NavBar";
import "./Spot.css";
import { BsDoorOpen } from "react-icons/bs";
import { BsCalendarDate } from "react-icons/bs";
import EditSpot from "./EditSpot";
import ReserveSpot from "./ReserveSpot";
import { GiVacuumCleaner } from "react-icons/gi";
import { FaAddressCard } from "react-icons/fa";
import Maps from "../Maps/Maps";
import Footer from "../Footer/Footer";
import Reviews from "../Reviews/Reviews";

const Spot = ({ spotsLoaded }) => {
  const user = useSelector((state) => state.session.user);
  const spot = useSelector((state) => state.spotReducer);
  const { spotId } = useParams();
  const loginModal = useSelector((state) => state.modals.loginModal);
  const signupModal = useSelector((state) => state.modals.signUpModal);
  const [googleMapsApiKey, setGoogleMapsApiKey] = useState(false);

  useEffect(() => {
    if (googleMapsApiKey) {
      return;
    }
    (async () => {
      const res = await fetch("/api/maps", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ key: "key" }),
      });

      const data = await res.json();

      setGoogleMapsApiKey(data.key);
    })();
  }, [googleMapsApiKey]);

  const isUserLoggedIn = () => {
    if (user === null) {
      return (
        <div className="logInDivSpot">
          <div className="logInOrSignUpTextContainerSpotPage">
            <div onClick={() => loginModal(true)} className="logInTextSpotPage">
              Log in
            </div>
            <div className="orTextSpotPage">Or</div>
            <div
              onClick={() => signupModal(true)}
              className="signUpTextSpotPage"
            >
              {" "}
              Sign Up
            </div>
          </div>
        </div>
      );
    } else {
      return (
        <ReserveSpot
          spotId={parseInt(spotId, 10)}
          totalOccupantsAllowed={spot[`${spotId}`]?.total_occupancy}
          price={spot[`${spotId}`]?.price_per_day}
        />
      );
    }
  };

  let petsAllowed;
  let hasAc;
  let hasTv;
  let hasWifi;

  if (spot[`${spotId}`]?.pets_allowed === "True") {
    petsAllowed = "Yes";
  } else {
    petsAllowed = "No";
  }

  if (spot[`${spotId}`]?.has_ac === "True") {
    hasAc = "Yes";
  } else {
    hasAc = "No";
  }

  if (spot[`${spotId}`]?.has_tv === "True") {
    hasTv = "Yes";
  } else {
    hasTv = "No";
  }

  if (spot[`${spotId}`]?.has_wifi === "True") {
    hasWifi = "Yes";
  } else {
    hasWifi = "No";
  }

  let spotDiv = !spot[`${spotId}`] ? null : (
    <div className="spotPageContainer">
      <div className="spotPageNavBArContainerParent">
        <NavBar spotPage={true} />
      </div>
      <div className="spotNameAndImageSection">
        <div id="spotNameSpotPage">{spot[`${spotId}`]?.name}</div>
        <div className="spotPageImagesSection">
          <div
            style={{
              width: "49%",
              height: "357px",
              backgroundImage: `url(${spot[`${spotId}`]?.images[0]})`,
              backgroundSize: "cover",
              backgroundRepeat: "no-repeat",
              backgroundPosition: "center",
              borderRadius: "12px",
            }}
          ></div>
          <div
            style={{
              width: "49%",
              height: "357px",
              backgroundImage: `url(${spot[`${spotId}`]?.images[1]})`,
              backgroundSize: "cover",
              backgroundRepeat: "no-repeat",
              backgroundPosition: "center",
              borderRadius: "12px",
            }}
          ></div>
        </div>
      </div>
      <div className="spotInfoAndReserveOrEditContainer">
        <div className="spotDescriptionAndInfoDiv">
          <div className="spotDetailsTitleAndStats">
            <h2>
              {spot[`${spotId}`]?.type} hosted by {spot[`${spotId}`]?.host_name}
            </h2>
            <div className="spotStatContainer">
              Total Bathrooms {spot[`${spotId}`]?.total_bathrooms} - Total
              Bedrooms {spot[`${spotId}`]?.total_bedrooms} - Total Occupants
              Allowed {spot[`${spotId}`]?.total_occupancy}
            </div>
            <div className="spotStatContainerPetsAllowed">
              <div>Pets Allowed: {petsAllowed}</div>
              <div>Has Ac: {hasAc}</div>
              <div>Has Wifi: {hasWifi}</div>
              <div>Has TV: {hasTv}</div>
            </div>
          </div>
          <div className="reservationAndCancelationTextContainer">
            <div className="doorAndSelfTextContainer">
              {" "}
              <BsDoorOpen id="doorIcon" /> Self Check In{" "}
            </div>
            <div className="checkInYouselfText">
              Check yourself by showing up.
            </div>
            <div className="calendarAndCancelationText">
              {" "}
              <BsCalendarDate id="calendarIcon" /> Free cancellation before
              start date.{" "}
            </div>
          </div>
          <div className="reservationAndCancelationTextContainer">
            <div className="doorAndSelfTextContainer">
              {" "}
              <FaAddressCard
                style={{ fontSize: "25px", marginRight: "15px" }}
              />{" "}
              Address: {spot[`${spotId}`]?.address}{" "}
            </div>
            <div className="calendarAndCancelationText">
              {" "}
              <GiVacuumCleaner
                style={{ fontSize: "25px", marginRight: "15px" }}
              />{" "}
              Don't worry about cleaning! Daytrip cleaning service on us.
            </div>
          </div>
          <div className="spotDescriptionContainer">
            {spot[`${spotId}`]?.description}
          </div>
        </div>
        <div className="editDeleteOrReserveSection">
          {user?.id === parseInt(spot[`${spotId}`]?.host_id, 10) ? (
            <EditSpot spot={spot} spotId={spotId} />
          ) : (
            isUserLoggedIn()
          )}
        </div>
      </div>
      <div id="locationParentContainer">
        <div id="locationHeader">Where you'll be</div>
      </div>
      <div id="mapsParentContainerSpotPage">
        <div id="mapsContainerSpotPage">
          <Maps
            apiKey={googleMapsApiKey}
            center={{
              lng: spot[`${spotId}`].longitude,
              lat: spot[`${spotId}`].latitude,
            }}
            singleSpot={true}
            zoom={12}
          />
        </div>
      </div>
      {spot[`${spotId}`].reviews.length > 0 && (
        <Reviews reviews={spot[`${spotId}`].reviews} />
      )}
    </div>
  );

  if (!spotsLoaded) {
    return null;
  }
  if (!spot[`${spotId}`]) {
    return null;
  }
  if (!googleMapsApiKey) {
    return null;
  }

  return (
    <>
      {spot[`${spotId}`] === undefined ? (
        <div>Spot Does not Exist</div>
      ) : (
        spotDiv
      )}
      {spot[`${spotId}`] !== undefined && <Footer />}
    </>
  );
};

export default Spot;
